docs: initial draft of requirement process (#377)
eclipse-score/score: docs/process/process_areas/requirements_engineering/requirements_concept.rst

Inputs
******

#. Stakeholders for the requirements?
#. Who needs which information?
#. Which Requirement Levels can we derive from that?
#. Which Attributes are required?
#. How do the different requirement levels correlate to each other?

Stakeholders for the requirements
=================================

#. :need:`Tech Lead Circle <RL_technical_lead>`

   * Define specification and content for the platform
   * Creation of a project timeline
   * Track project progress

#. :need:`SW Architect <RL_committer>`

   * Break down the platform specification into features (High Level)
   * Derive component architecture for each feature
   * Allocate requirements to architecture elements for specification of features/components
   * Define AoUs which arise from architecture

#. :need:`Tester <RL_committer>`

   * Verify that the specification is fulfilled by the elements under test
   * Consider AoUs for test case specification

> #. :need:`Safety Architect <RL_safety_manager>`

Reviewer: Security Architect, 
Trust Boundary Analysis, Defense in Depth Analysis
Qualitative Security Analysis (TARA or at least Attack Potential Analysis with impact category Safety)

Author: added

   * *Dependent Failure Analysis*

     * Requires inputs towards independence and interference of the element under investigation

   * *Qualitative safety analysis* (e.g. FMEA)

     * Detailed element description to identify systematic errors within the element under investigation

#. :need:`Platform/Tooling SW Developer <RL_committer>`

   * Implement the SW according to its specification
   * Create traceability by linking its specification to code
   * Consider AoUs of other components
   * Create AoUs for component under development

#. Feature User

   * Get detailed information concerning the specification of a feature
   * Be informed about its boundary conditions (AoUs)

#. :need:`Platform SW Developer of the Reference Integration <RL_committer>`

   * Requirements for Integration

ISO26262 Requirements
=====================

.. needtable:: Overview of SWE1 ASPICE Requirements
   :style: table
   :columns: title;id
   :filter: search("(STD_REQ_ISO26262__support_[1-9]$)", id) or "STD_REQ_ISO26262__support_10" in id
   :colwidths: 70,30

ASPICE Requirements
===================

> .. needtable:: Overview of SWE1 ASPICE Requirements

Reviewer: ASPICE standard is now available in another PR

Author: it was decided to kick out overview of standard requirements

   :style: table
   :columns: title;id
   :filter: "SWE-1-BP" in id
   :colwidths: 70,30

Requirement Levels
******************

> Based on the inputs of the previous chapter the types of requirements which need to be implemented in the project can be derived. The defined levels are shown in the *Building Blocks Meta Model* <TBD>

Reviewer: You can link also to General Concept, another PR

Author: No link there yet

Stakeholder Requirements
========================

On the platform level the *Stakeholder (=customer) Requirements* are defined. These requirements describe which content the platform needs to contain, and serve as a project description of the top-level functionality. An example could be e.g.

.. code-block:: text

   The platform shall support configuration of applications via files (e.g. yaml, json)

Feature Requirements
====================

> The next level of requirements addresses mainly the integration level of SW modules and components. On this level the behavior of the feature on platform level shall be described including the correlations of the integrated components. It serves mainly as an input for (SW + Safety) Architects, Testers, Integrators. Those so called *Feature Requirements* are derived from the *Stakeholder Requirements*. To provide an example

Reviewer 2: ```suggestion
The *Feature Requirements* addresses mainly the integration level of SW modules and components. These shall describe the behavior of the feature on platform level including the correlations of the integrated components. These serves mainly as an input for (SW + Safety) Architects, Testers, Integrators. Those so called *Feature Requirements* are derived from the *Stakeholder Requirements*. To provide an example
```

.. code-block:: text

   The feature shall accept JSON formatted string input according to RFC-8259

However the detailed interaction of the underlying components itself which is required to form a feature shall be defined in the feature architecture.

Component Requirements
======================

The lowest abstraction level is represented by the *component requirements*. They are derived from *feature requirements* and describe component specific implementation details. It is described which behavior a component itself needs to fulfil in the context of the feature, e.g.

.. code-block:: text

   The component shall provide API calls to read and interpret every field of a JSON body in C++

> AoU Requirements

Reviewer 2: ```suggestion
Assumption of Use Requirements
```

Author: changed

Process Requirements
====================

Besides those four types of requirements which describe the contents of the platform also a type, describing the requirements towards the tooling from a process point of view, needs to be specified. These *process requirements* can be derived from a process description. Here it is defined which part of the process need to be performed manually and which parts of the process should be implemented by tooling

.. code-block:: text

   It shall be checked that safety requirements (Safety != QM) can only be linked against safety requirements.

.. _attributes of the requirements:

Attributes of the Requirements
******************************

The required attributes for the requirements are defined in this subchapter. On the top level we can distinguish between attributes which need to be filled manually and attributes which are generated during the *docs build*.

Following attributes need to be filled manually for each requirement:

.. list-table:: Manual Attributes
   :header-rows: 1
   :widths: 15,85

   * - Attribute
     - Description
   * - Status

>      - The status for a requirement can either be valid or invalid. The reasoning for this is that the goal is to only have valid requirements in the main branch. So a status *draft* is not required. It is obvious that the requirement is in the status draft as long as the PR is not merged.

Reviewer: Is valid, invalid still in sync with our metamodel?

Author: need to discuss it in detail if we still want "invalid"
currently it is still in synch with the metamodel which is defined in the yaml file

     - It contains the hash of the parent requirement. If the parent requirement is changed the hash will also change and the linkage has to be revisited again. A more detailed description is provided here: :need:`GD_REQ__req__attr_hash`
     - Script / Bazel Target
   * - Implemented by
     - During Build the code files are parsed for a defined tag which includes the requirement id. If this is located a link to the code will be added in the requirement
     - Sphinx Needs Build
   * - Verified by
     - During build the junit test files are parsed for a defined marker which includes the requirement id. If the marker is located in the test a link to the test case will be added to the requirement
     - Sphinx Needs Build
   * - Requirement Covered
     - During build it will be checked if the requirements hashes which are mentioned in the coverage file match the hashes of the linked child requirements. If so then this attribute will be set to yes.
     - Sphinx Needs Build

More details about the generation of the automated attributes are explained in the following chapter where the general workflow for generating requirements including their status is shown.

> For creating requirements templates are available for each specific type: :ref:`requirement templates`

Author: Remove Link to Templates, not required for concept

Author: removed

During Sphinx build it shall be checked if the attribute hash matches the actual has value of the requirement.

Linking child requirements including hashes
===========================================

If a requirement is linked to a top level requirement also the hash of the target requirement shall be part of the link. It shall automatically be written into the attribute *satisfies hash*. Upon Sphinx Build it shall be checked if the attribute *satisfies hash* matches the calculated hash of the requirement which is linked via *satisfies*.

As this check is included in the sphinx build as a warning it can be guaranteed that a change of a parent requirement can only be merged if the `linkhashes` in the requirements are also updated in a `Depends-On` PR.

.. figure:: _assets/requirements_versioning.svg
   :alt: Requirements Versioning
   :align: center
   :width: 50%

   Versioning of Requirements

.. _reviews of the requirements:

Reviews of the Requirements
***************************

> Some of the checks cannot be performed automatically. Therefore a manual inspection of the requirements is needed. The requirement review itself is triggered when a contributor wants to trigger a requirement review. For this review a checklist is available: :need:`GD_CHKLST__req__inspection`.

Author: Remove Link to Templates + Add Link to General Review Guide

Author: Removed + Text added

If this file will now be merged to the main branch a review will be triggered again. During this review it will only be checked if the parent requirement is covered by its child requirements.

Additionally during build it shall be checked if exactly the requirements and hashes which are specified in this file are linked to the requirement in the current documentation. If all items match then the requirement can be seen as covered. If now a child requirement is changed its hash will also be changed and the requirement coverage needs to be revisited again.

.. _traceability concept for requirements:

Traceability Concept for Requirements
*************************************

The standards require that a requirement can be traced throughout the complete hierarchy levels including its implementation and verification <TBD: Link>. In this project it is implemented the following way:

> In general the traceability is visualized is in main development work product traceability model (:numref:`wp_traceability_model`).

Reviewer 2: ```suggestion
In general the traceability is visualized in main development work product traceability model (:numref:`wp_traceability_model`).
```

Author: fixed
